### ISSUE TITLE
Presence of `default` value for an object property should not imply `required`.

### ISSUE THREAD

**Issue by nhollander-alert** [ThomasAribart/json-schema-to-ts#211]
According to the json schema specification, the `default` keyword is a meta-data annotation for "documentation and user interface display purposes" [^1]. The schema reference states that "...`default` is typically used to express that if a value is missing, then the value is semantically the same as if the value was present with the default value." [^2]

Based on this, I would expect that the `default` value would be entirely ignored by this library, as it is a purely semantic marker which tells users that an `undefined` value will be treated by the tool as if it were the given value, but there is no requirement that processing tools substitute `undefined` for the default value.

Given the following schema definition this library correctly produces an object where `foo` is required but `bar` is optional.

```ts
type test = FromSchema<{
    type: "object",
    additionalProperties: false,
    required: ["foo"],
    properties: {
        foo: { type: "string" },
        bar: { type: "string" }
    }
}>;
```
```ts
type test = {
    bar?: string | undefined;
    foo: string;
}
```

If we alter `bar` to include a `default` value, the resulting type no longer accepts `undefined` as a valid value, despite it still being an allowed value per the spec.
```ts
type test = FromSchema<{
    type: "object",
    additionalProperties: false,
    required: ["foo"],
    properties: {
        foo: { type: "string" },
        bar: { type: "string", default: "abcd" }
    }
}>;
```
```ts
type test = {
    foo: string;
    bar: string;
}
```

[^1]: https://json-schema.org/draft/2020-12/json-schema-validation#section-9.2
[^2]: https://json-schema.org/understanding-json-schema/reference/annotations

**Comment by Caesar2011** [comment]
It depends on the perspective. If you have an API request following this schema and you then use the data, it will be present. If you are on the other side and you need to fulfil a given API spec (JSON schema), then it sould be optional.

**Comment by nhollander-alert** [comment]
While APIs are _permitted_ to apply `default` values to incoming or outgoing requests, it is not mandated by the specification and APIs are free to send or omit default values. Unless you pass your incoming or outgoing data through some kind of function to apply default values, it is not safe to assume that a value will ever be defined just because it has a `default` field.

**Comment by ThomasAribart** [comment]
@nhollander-alert Good point. There's already a `keepDefaultedPropertiesOptional` option to revert this, but the default is `false`.

I could reverse it and rename it `requireDefaultedProperties`. That would be a breaking change but there's not been any breaking change for a while so I'm actually fine with it.
